lean4/src/putnam_2023_b3.lean

```diff
@@ -0,0 +1,20 @@
+import Mathlib
+
+open ProbabilityTheory Set
+
+noncomputable abbrev putnam_2023_b3_solution : ℕ → ℝ := sorry
+-- (fun n : ℕ ↦ (2 * (n : ℝ) + 2) / 3)
+/--
+A sequence $y_1, y_2, \ldots, y_k$ of real numbers is called zigzag if $k = 1$, or if $y_2 - y_1, y_3 - y_2, \ldots, y_k - y_{k-1}$ are nonzero and alternate in sign. Let $X_1, X_2,\ldots, X_n$ be chosen independently from the uniform distribution on $[0,1]$. Let $a(X_1, X_2, \ldots, X_n)$ be the largest value of $k$ for which there exists an increasing sequence of integers $i_1, i_2, \ldots, i_k$ such that $X_{i_1}, X_{i_2}, \ldots, X_{i_k}$ is zigzag. Find the expected value of $a(X_1, X_2, \ldots, X_n)$ for $n \ge 2$.
+-/
+theorem putnam_2023_b3
+    (IsZigZag : {k : ℕ} → (Fin k → ℝ) → Prop)
+    (IsZigZag_def : ∀ (k : ℕ) [NeZero k] (y : Fin k → ℝ),
+      IsZigZag y ↔ k = 1 ∨ ((∀ i, i + 1 < k → y (i + 1) ≠ y i)) ∧
+        (∀ i, i + 2 < k → y (i + 2) < y (i + 1) ↔ y i < y (i + 1)))
+    (n : ℕ)
+    (hn : 2 ≤ n)
+    (a : (Fin n → Icc (0 : ℝ) 1) → ℕ)
+    (ha : ∀ x, IsGreatest {k | ∃ i : Fin k ↪o Fin n, IsZigZag ((↑) ∘ x ∘ i)} (a x)) :
+    𝔼[(↑) ∘ a] = putnam_2023_b3_solution n := by
```

Comment:
Clarification Q: The default probability measure on [0,1]^n in Lean is the lebesgue measure, right?

Reply:
Yes, this picks up the product measure of the Lebesgue measure. You can investigate this sort of thing as follows:
```lean
variable (n : ℕ)
#synth MeasureTheory.MeasureSpace (Fin n → Icc (0 : ℝ) 1) -- Will reply `MeasureTheory.MeasureSpace.pi`
```
and from there, look at the definition of `MeasureTheory.MeasureSpace.pi`.

Reply:
Having said that, it's better to look at API than definitions so this is really the right evidence:
```lean
variable (n : ℕ)

open MeasureTheory

example (s : Fin n → Set (Icc (0 : ℝ) 1)) :
    volume (univ.pi s) = ∏ i, volume (s i) :=
  volume_pi_pi s
```

Reply:
I guess I should also give you evidence that the measure on $[0, 1]$ is the restriction of the Lebesgue measure. Here's a rather ugly proof:
```lean
open MeasureTheory

example (a b : Icc (0 : ℝ) 1) (hab : a ≤ b) :
    (volume (Icc a b)).toReal = (b : ℝ) - (a : ℝ) := by
  have foo : (volume : Measure (Icc (0 : ℝ) 1)) =
    (volume : Measure ℝ).comap ((↑) : (Icc (0 : ℝ) 1) → ℝ) := rfl
  have bar : MeasurableEmbedding ((↑) : (Icc (0 : ℝ) 1) → ℝ) :=
    MeasurableEmbedding.subtype_coe measurableSet_Icc
  rw [foo, Measure.comap_apply _ Subtype.val_injective
    (fun s hs ↦ bar.measurableSet_image.mpr hs) _ measurableSet_Icc]
  simpa
```

Reply:
Thank you!